File: Python/LCA.py
from tree import Node

def lowest_common_ancestor_helper(node1, node2, currentNode):
    if node1==node2:
        return 2, node1
    numFound = 0
    if currentNode.left:
        lNumFound,lNode = lowest_common_ancestor_helper(node1,node2,currentNode.left)
        if lNumFound == 2:
            return 2, lNode
        numFound += lNumFound
    if currentNode.right:
        rNumFound,rNode = lowest_common_ancestor_helper(node1,node2,currentNode.right)
        if rNumFound == 2:
            return 2, rNode
        numFound += rNumFound
    if currentNode == node1 or currentNode == node2:
        numFound += 1
    #print (numFound, currentNode.data)
    return numFound, currentNode

def lowest_common_ancestor(node1, node2, root):
    numFound, node = lowest_common_ancestor_helper(node1, node2, root)
    return node.data





def test_tree_creation():
    root = Node("root")
    right = Node("right")
    left = Node("left")
    lr = Node("lr")
    ll = Node("ll")
    root.right = right
    root.left = left
    left.right = lr
    left.left = ll

    print("\n")
    root.print_subtrees()
    print("\n")

    print(lowest_common_ancestor(ll, ll, root))

File: Ruby/lca.rb
#first attempt w ruby :-)
class Node
  attr_accessor :left
  attr_accessor :right

  def initialize(data)
    @data = data
  end

  def print_data()
    puts @data
  end

  def print_subtrees(indentation = "")
    if  !@right.nil?
      @right.print_subtrees(indentation + "\t")
    end
    if !@left.nil? || !@right.nil?
      puts indentation+@data+" ----"
    else
      puts indentation+@data
    end
    if  !@left.nil?
      @left.print_subtrees(indentation + "\t")
    end
    
  end
end



def lowest_common_ancestor (node1, node2, currentNode)
  if node1 == node2
    #puts node1.print_data 
    return currentNode.print_data
  else
    found = 0
    if !currentNode.left.nil?
      leftfound = lowest_common_ancestor(node1, node2, currentNode.left)
      if leftfound.instance_of? Fixnum
        found += leftfound
      else
        return leftfound
      end
    end

    if !currentNode.right.nil?
      rightfound = lowest_common_ancestor(node1, node2, currentNode.right)
      if rightfound.instance_of? Fixnum
        found += rightfound
      else
        return rightfound
      end
    end

    if rightfound == 2 || leftfound == 2
      return 2
    end

    if currentNode == node1 || currentNode == node2
      found += 1
    end

    if found == 2
      #puts currentNode.print_data
      return currentNode.print_data
    end

    return found
  end
end





# root = Node.new("Root")
# right = Node.new("R")
# left = Node.new("L")
# leftright = Node.new("LR")
# leftleft = Node.new("LL")
# leftleftright = Node.new("LLR")
# rightleft = Node.new("RL")

# root.left = left
# root.right = right
# left.right = leftright
# left.left = leftleft
# leftleft.right = leftleftright
# right.left = rightleft

# puts "\n"
# root.print_subtrees
